File: backend/majors.py
"""Career major → ordered track (Language slug) definitions, mirroring the
frontend's src/majors.js. Keeping the mapping here lets the Roadmap API resolve
a major to its ordered learning path server-side.
"""

MAJOR_TRACKS = {
    "computer-science": ["python", "python-intermediate", "dsa", "javascript", "linux-shell"],
    "data-science": ["python", "python-intermediate", "sql-data", "backend-foundations", "linux-shell"],
    "data-analyst": ["python", "python-intermediate", "sql-data", "linux-shell"],
    "ai-engineer": ["python", "python-intermediate", "sql-data", "backend-foundations", "linux-shell", "ai-llm"],
    "web-developer": ["html-css", "javascript", "react-typescript", "full-stack"],
    "backend-engineer": ["python", "python-intermediate", "sql-data", "backend-foundations", "linux-shell", "full-stack"],
    "automation": ["python", "python-intermediate", "linux-shell", "backend-foundations"],
    # NOTE: network-engineer reuses existing tracks until a dedicated
    # `networking` track (OSI/TCP-IP/DNS/HTTP, network automation) is seeded.
    "network-engineer": ["linux-shell", "python", "python-intermediate", "backend-foundations"],
}

# Library shelves each major should see first, most relevant first. Slugs point at
# DocCollection.slug — mirrored track slugs (same as MAJOR_TRACKS) plus the
# standalone reference shelves. Mirror the frontend's src/libraryMap.js.
_STANDALONE_DOCS = ["version-control", "dev-workflow"]
MAJOR_DOCS = {
    major: tracks + _STANDALONE_DOCS
    for major, tracks in MAJOR_TRACKS.items()
}

# Human labels, mirroring the frontend's src/majors.js.
MAJOR_LABELS = {
    "computer-science": "Computer Science",
    "data-science": "Data Science",
    "data-analyst": "Data Analyst",
    "ai-engineer": "AI Engineer",
    "web-developer": "Web Developer",
    "backend-engineer": "Backend Engineer",
    "automation": "Automation Engineer",
    "network-engineer": "Network Engineer",
}

# Skill buckets for the Progress "skills" view and Career readiness. Each bucket
# aggregates progress from lesson tracks (Language.slug), challenge languages, and
# challenge topics. Keys are stable identifiers used by MAJOR_SKILLS below.
SKILL_DEFS = [
    {
        "key": "python",
        "label": "Python",
        "tracks": ["python", "python-intermediate"],
        "challenge_langs": ["python"],
        "challenge_topics": [],
    },
    {
        "key": "javascript",
        "label": "JavaScript",
        "tracks": ["javascript"],
        "challenge_langs": ["javascript"],
        "challenge_topics": ["javascript"],
    },
    {
        "key": "web",
        "label": "Web & Frontend",
        "tracks": ["html-css", "react-typescript", "full-stack"],
        "challenge_langs": [],
        "challenge_topics": [],
    },
    {
        "key": "backend",
        "label": "Backend & APIs",
        "tracks": ["backend-foundations", "full-stack"],
        "challenge_langs": [],
        "challenge_topics": [],
    },
    {
        "key": "sql",
        "label": "SQL & Data",
        "tracks": ["sql-data"],
        "challenge_langs": ["sql"],
        "challenge_topics": ["sql"],
    },
    {
        "key": "shell",
        "label": "Linux & Shell",
        "tracks": ["linux-shell"],
        "challenge_langs": [],
        "challenge_topics": [],
    },
    {
        "key": "problem-solving",
        "label": "Problem Solving",
        "tracks": [],
        "challenge_langs": [],
        "challenge_topics": [
            "algorithms", "arrays", "strings", "recursion",
            "stacks", "dictionaries", "control-flow",
        ],
    },
    {
        "key": "dsa",
        "label": "Data Structures & Algorithms",
        "tracks": ["dsa"],
        "challenge_langs": [],
        "challenge_topics": [
            "algorithms", "arrays", "recursion", "stacks", "dictionaries",
        ],
    },
    {
        "key": "ai-llm",
        "label": "AI & LLM Apps",
        "tracks": ["ai-llm"],
        "challenge_langs": [],
        "challenge_topics": ["ai", "llm", "prompting", "rag"],
    },
]

SKILL_LABELS = {s["key"]: s["label"] for s in SKILL_DEFS}

# The skills that define "job ready" for each major, most central first.
MAJOR_SKILLS = {
    "computer-science": ["dsa", "python", "problem-solving", "javascript", "shell"],
    "data-science": ["python", "sql", "problem-solving"],
    "data-analyst": ["sql", "python", "problem-solving"],
    "ai-engineer": ["ai-llm", "python", "problem-solving", "sql", "backend"],
    "web-developer": ["javascript", "web", "problem-solving"],
    "backend-engineer": ["python", "backend", "sql", "shell", "problem-solving"],
    "automation": ["python", "shell", "backend"],
    "network-engineer": ["shell", "python", "backend"],
}

File: frontend/src/majors.js
// Career "majors". Picking one filters + orders the dashboard tracks and — most
// importantly — steers the AI project generator via `projectFocus`.
//
// `tracks` holds Language slugs (from the lessons API) this major covers, most
// relevant first. `projectLang` is the language the project generator defaults
// to. `icon` is a Tabler icon name resolved in MajorIcon.

export const MAJORS = {
  'computer-science': {
    label: 'Computer Science',
    icon: 'TbBinaryTree',
    color: '#8B5CF6',
    tagline: 'Think like a computer scientist',
    blurb:
      'The fundamentals every engineer shares: logic, data structures, algorithms and how to reason about efficiency.',
    focus: ['Programming fundamentals', 'Data structures', 'Algorithms', 'Recursion', 'Big-O thinking', 'OOP'],
    tracks: ['python', 'python-intermediate', 'dsa', 'javascript', 'linux-shell'],
    projectLang: 'python',
    projectFocus:
      'core computer-science fundamentals — implementing data structures, algorithms and small command-line programs that exercise logic, recursion and complexity analysis',
  },

  'data-science': {
    label: 'Data Science',
    icon: 'TbChartHistogram',
    color: '#4ADE80',
    tagline: 'Turn raw data into answers',
    blurb:
      'Load messy data, clean it, measure it and communicate what you found. Python plus SQL plus a statistical mindset.',
    focus: ['Python for data', 'Working with datasets', 'Statistics basics', 'SQL queries', 'Data cleaning', 'Reporting'],
    tracks: ['python', 'python-intermediate', 'sql-data', 'backend-foundations', 'linux-shell'],
    projectLang: 'python',
    projectFocus:
      'data analysis — loading a dataset, cleaning it, computing summary statistics and metrics, and producing a clear text or tabular report of the findings',
  },

  'data-analyst': {
    label: 'Data Analyst',
    icon: 'TbChartHistogram',
    color: '#34D399',
    tagline: 'Answer business questions with data',
    blurb:
      'Query databases directly, shape and aggregate the results, and turn them into clear answers. SQL-first, with just enough Python.',
    focus: ['SQL queries', 'Joins & aggregation', 'Window functions', 'Python for data', 'Data cleaning', 'Reporting'],
    tracks: ['python', 'python-intermediate', 'sql-data', 'linux-shell'],
    projectLang: 'python',
    projectFocus:
      'data analysis — writing SQL to pull and aggregate data, loading results in Python, computing metrics and summary tables, and producing a clear written report of the findings',
  },

  'ai-engineer': {
    label: 'AI Engineer',
    icon: 'TbSparkles',
    color: '#22D3EE',
    tagline: 'Build with models',
    blurb:
      'The practical side of shipping AI features: model APIs, structured output, tool use, retrieval (RAG), evaluation and safe serving — on a Python + backend base.',
    focus: ['Python', 'Model APIs & tool use', 'Prompt design', 'Structured output', 'Embeddings & RAG', 'Evaluation & guardrails'],
    tracks: ['python', 'python-intermediate', 'sql-data', 'backend-foundations', 'linux-shell', 'ai-llm'],
    projectLang: 'python',
    projectFocus:
      'AI engineering — building LLM-powered features: a prompt-driven assistant with structured JSON output, a tool-calling agent loop, a small RAG pipeline over local documents, or an evaluation harness that scores model output against golden answers',
  },

  'web-developer': {
    label: 'Web Developer',
    icon: 'TbBrandHtml5',
    color: '#FB923C',
    tagline: 'Make things people click',
    blurb:
      'Structure with HTML, style with CSS, bring it to life with JavaScript. Interfaces, interactivity and the browser.',
    focus: ['HTML', 'CSS', 'JavaScript', 'DOM & events', 'Fetch / HTTP', 'Components'],
    tracks: ['html-css', 'javascript', 'react-typescript', 'full-stack'],
    projectLang: 'javascript',
    projectFocus:
      'front-end web development — building interactive pages and UI components with HTML, CSS and JavaScript: forms, lists, filtering, DOM updates and calling an API for data',
  },

  'backend-engineer': {
    label: 'Backend Engineer',
    icon: 'TbServer2',
    color: '#3B82F6',
    tagline: 'Power the app from behind',
    blurb:
      'APIs, databases, migrations and the Git workflow. The services and data models everything else depends on.',
    focus: ['Python', 'REST APIs', 'Databases & SQL', 'Schema migrations', 'Validation & auth', 'Git & GitHub'],
    tracks: ['python', 'python-intermediate', 'sql-data', 'backend-foundations', 'linux-shell', 'full-stack'],
    projectLang: 'python',
    projectFocus:
      'back-end engineering — designing REST endpoints, request/response models, validation, and small services that read and write structured data (in-memory or SQL)',
  },

  automation: {
    label: 'Automation Engineer',
    icon: 'TbScript',
    color: '#2DD4BF',
    tagline: 'Let scripts do the boring parts',
    blurb:
      'Practical Python that processes files, transforms text, talks to APIs and removes repetitive manual work.',
    focus: ['Python scripting', 'File I/O', 'Text & regex', 'Calling web APIs', 'Error handling', 'Scheduling'],
    tracks: ['python', 'python-intermediate', 'linux-shell', 'backend-foundations'],
    projectLang: 'python',
    projectFocus:
      'practical automation — scripts that read and write files, parse or transform text, call a web API and glue steps together to automate a repetitive task',
  },

  // NOTE: reuses shared tracks until a dedicated `networking` track
  // (OSI / TCP-IP / DNS / HTTP, then network automation) is added.
  'network-engineer': {
    label: 'Network Engineer',
    icon: 'TbNetwork',
    color: '#38BDF8',
    tagline: 'Keep the packets moving',
    blurb:
      'The Linux command line, scripting and services behind networked systems — a base for TCP/IP, DNS, HTTP and network automation.',
    focus: ['Linux & shell', 'Processes & permissions', 'Python scripting', 'HTTP & APIs', 'Automation', 'Troubleshooting'],
    tracks: ['linux-shell', 'python', 'python-intermediate', 'backend-foundations'],
    projectLang: 'python',
    projectFocus:
      'network / infrastructure automation — Python scripts that check connectivity, parse logs or config, call device or cloud APIs, and report status on a schedule',
  },
};

export const MAJOR_KEYS = Object.keys(MAJORS);
export const MAJOR_STORAGE_KEY = 'cs-major';

export function isMajor(key) {
  return MAJOR_KEYS.includes(key);
}
